### PR TITLE
Engine - Add helper to filter and map using definition variables

### PR DESCRIPTION
**Pull request by nbertoldo**
|Related issue|
|---|
|#17027|

## Description

In the current implementation, kvdb filter and map helpers are designed for large datasets. However, there is a need to adapt this functionality for smaller datasets.

Now that we can define definitions with any JSON value, we want to implement new helpers using these definitions as the datasets.

The helpers will accept a definition variable (array or object) as an argument, implementing the following functionality:

- [x]  `+match_value`: Filter that target field value is present in the specified definition array.
- [x]  `+match_key`: Filter that target field value is present as a key in the specified definition object.
- [x]  `+get_value`: Map target field value with the content of the some key in the definition object, where the key is specified with a reference to another field.

### +match_value
Decoder:
```
name: decoder/test/0

definitions:
  ~defArray: [2, 4, 6]

parse:
  logpar:
    - wazuh.message: <~json/json>

normalize:
  - check:
    - ~json.field1: +match_value/$~defArray
    - ~json.field2: +match_value/$~json.refArray
```
Output:
```
Enter a log in single line (Crtl+C to exit):

{"refArray": ["value1", "value2", "value3"], "field1": 4, "field2": "value3"}


DECODERS:
  decoder/test/0  ->  success
    [decoder/test/0] [/wazuh/message: <~json/json>] -> Success
    [decoder/test/0] [condition]:success
    [decoder/test/0] [helper.match_value[/~0json/field1, [2,4,6]]] -> Success
    [decoder/test/0] [helper.match_value[/~0json/field2, /~0json/refArray]] -> Success

RULES:


OUTPUT:

{
    "wazuh": {
        "queue": 49,
        "location": "/dev/stdin",
        "message": "{\"refArray\": [\"value1\", \"value2\", \"value3\"], \"field1\": 4, \"field2\": \"value3\"}"
    },
    "~json": {
        "refArray": [
            "value1",
            "value2",
            "value3"
        ],
        "field1": 4,
        "field2": "value3"
    }
}
```
### +match_key
Decoder:

```
name: decoder/test/0

definitions:
  ~defObject: {"key1": "value1", "key2": 10}

parse:
  logpar:
    - wazuh.message: <~json/json>

normalize:
  - check:
    - ~json.field1: +match_key/$~defObject
    - ~json.field2: +match_key/$~json.refObject
    - ~json.field3: +match_key/$~defObject
```
Output:
```
Enter a log in single line (Crtl+C to exit):

{"refObject": {"key1": 49, "key2": "hello"}, "field1": "key2", "field2": "key1", "field3": "keyX"}


DECODERS:
  decoder/test/0  ->  success
    [decoder/test/0] [/wazuh/message: <~json/json>] -> Success
    [decoder/test/0] [condition]:success
    [decoder/test/0] [helper.match_key[/~0json/field1, {"key1":"value1","key2":10}]] -> Success
    [decoder/test/0] [helper.match_key[/~0json/field2, /~0json/refObject]] -> Success
    [decoder/test/0] [helper.match_key[/~0json/field3, {"key1":"value1","key2":10}]] -> Failure: Object '{"key1":"value1","key2":10}' does not contain '/~0json/field3'

RULES:


OUTPUT:

{
    "wazuh": {
        "queue": 49,
        "location": "/dev/stdin",
        "message": "{\"refObject\": {\"key1\": 49, \"key2\": \"hello\"}, \"field1\": \"key2\", \"field2\": \"key1\", \"field3\": \"keyX\"}"
    },
    "~json": {
        "refObject": {
            "key1": 49,
            "key2": "hello"
        },
        "field1": "key2",
        "field2": "key1",
        "field3": "keyX"
    }
}
```
### +get_value
Decoder:
```
name: decoder/test/0

definitions:
  ~defObject: {"key1": "value1", "key2": 10}

parse:
  logpar:
    - wazuh.message: <~json/json>

normalize:
  - map:
    - ~json.field1: +get_value/$~defObject/$~json.key1
    - ~json.field2: +get_value/$~json.refObject/$~json.key2
    - ~json.field3: +get_value/$~json.defObject/$~json.key3
```
Output:
```
Enter a log in single line (Crtl+C to exit):

{"refObject": {"key1": 49, "key2": "hello"}, "key1": "key1", "key2": "key2", "key3": "key3"}


DECODERS:
  decoder/test/0  ->  success
    [decoder/test/0] [/wazuh/message: <~json/json>] -> Success
    [decoder/test/0] [condition]:success
    [decoder/test/0] [helper.get_value[/~0json/field1, {"key1":"value1","key2":10}, /~0json/key1]] -> Success
    [decoder/test/0] [helper.get_value[/~0json/field2, /~0json/refObject, /~0json/key2]] -> Success
    [decoder/test/0] [helper.get_value[/~0json/field3, /~0json/defObject, /~0json/key3]] -> Failure: Parameter '/~0json/defObject' not found

RULES:


OUTPUT:

{
    "wazuh": {
        "queue": 49,
        "location": "/dev/stdin",
        "message": "{\"refObject\": {\"key1\": 49, \"key2\": \"hello\"}, \"key1\": \"key1\", \"key2\": \"key2\", \"key3\": \"key3\"}"
    },
    "~json": {
        "refObject": {
            "key1": 49,
            "key2": "hello"
        },
        "key1": "key1",
        "key2": "key2",
        "key3": "key3",
        "field1": "value1",
        "field2": "hello"
    }
}
```

### CODE REVIEW

**Review by JavierBejMen — APPROVED** [review]
LGTM

**Review by JavierBejMen — APPROVED** [review]
LGTM

File: src/engine/source/builder/builders/opBuilderHelperFilter.cpp [hunk 1]
```diff
@@ -1479,4 +1479,254 @@ base::Expression opBuilderHelperIsFalse(const std::string& targetField,
...
+    const std::string failureTrace4 {
+        fmt::format("[{}] -> Failure: Parameter '{}' is not an array", name, parameters[0].m_value)};
+    const std::string failureTrace5 {
+        fmt::format("[{}] -> Failure: Array '{}' does not contain '{}'", name, parameters[0].m_value, targetField)};
+
+    // Return Term
+    return base::Term<base::EngineOp>::create(
+        name,
+        [=,
+         targetField = std::move(targetField),
```
Comment (juliancnn) [inline src/engine/source/builder/builders/opBuilderHelperFilter.cpp:1539]: why copy the traces string but move the targetField string?
I think we should unify

**Review by juliancnn — APPROVED** [review]
LGTM, it is only a feedback, not a request for changes.

**Review by LucioDonda — APPROVED** [review]
LGTM! 
if it's not too much additional development It should be an excellent addition to include this helper in a benchmark test. To have an idea of the cost of its use with large arrays.

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Pull request merged**